File: src/02_algorithms/01_sorting/03_insertion_sort/implementations/go/insertion_sort.go
package insertionSort

func InsertionSort(array []int) {
	for i := 1; i < len(array); i++ {
		key := array[i]
		j := i

		for j > 0 && array[j-1] > key {
			array[j] = array[j-1]
			j--
		}

		array[j] = key
	}

}

File: src/02_algorithms/01_sorting/03_insertion_sort/implementations/ts/insertion_sort.ts
export const insertion_sort = (array: Array<number>) => {
  // Iterate from second item to last
  for (let i = 1; i < array.length; i++) {
    const key = array[i]; // Value of current key position, start in second element
    let j = i;

    while (j > 0 && array[j - 1] > key) {
      array[j] = array[j - 1]; // Replace second element with first one
      j--; // Decrease counter, so we perform comparison and replace previous element if necessary
    }

    array[j] = key; // End of outer loop, insert key
  }
};
